gnovm/pkg/gnolang/dap_handlers.go

```diff
@@ -0,0 +1,837 @@
+		SupportsStepBack:                   false,
+		SupportsSetVariable:                false,
+		SupportsRestartFrame:               false,
+		SupportsGotoTargetsRequest:         false,
+		SupportsStepInTargetsRequest:       false,
+		SupportsCompletionsRequest:         false,
+		SupportsModulesRequest:             false,
+		SupportsRestartRequest:             false,
+		SupportsExceptionOptions:           false,
+		SupportsValueFormattingOptions:     false,
+		SupportsExceptionInfoRequest:       false,
+		SupportTerminateDebuggee:           true,
+		SupportsDelayedStackTraceLoading:   false,
+		SupportsLoadedSourcesRequest:       false,
+		SupportsLogPoints:                  false,
+		SupportsTerminateThreadsRequest:    false,
+		SupportsSetExpression:              false,
+		SupportsTerminateRequest:           true,
+		SupportsDataBreakpoints:            false,
+		SupportsReadMemoryRequest:          false,
+		SupportsDisassembleRequest:         false,
+		SupportsCancelRequest:              false,
+		SupportsBreakpointLocationsRequest: false,
+	}
+
+	s.initialized = true
+	return s.sendMessage(resp)
+}
+
+// handleLaunch processes the launch request
+func (s *DAPServer) handleLaunch(req *Request, _ []byte) error {
+	var args LaunchArguments
+	if err := json.Unmarshal(req.Arguments, &args); err != nil {
+		return err
+	}
+
+	// Send response
+	resp := NewResponse(req, true)
+	if err := s.sendMessage(resp); err != nil {
+		return err
+	}
+
+	// Send initialized event
+	event := &InitializedEvent{
+		Event: *NewEvent("initialized"),
+	}
+	return s.sendMessage(event)
+}
+
+// handleAttach processes the attach request
+func (s *DAPServer) handleAttach(req *Request, _ []byte) error {
+	var args AttachArguments
+	if err := json.Unmarshal(req.Arguments, &args); err != nil {
+		return err
+	}
+
+	// For Gno debugger, attach mode means the program was started with --attach flag
+	// and is waiting for debugger to connect before starting execution.
+
+	// If we're in attach mode, start the program execution now
+	if s.attachMode && s.programFiles != nil {
+		go func() {
+			// Run the loaded files
+			s.machine.RunFiles(s.programFiles...)
+			// Run main expression
+			if s.mainExpr != "" {
+				ex, err := ParseExpr(s.mainExpr)
+				if err == nil {
+					s.machine.Eval(ex)
+				}
+			}
+
+			// Send terminated event when done
+			s.SendTerminatedEvent()
+		}()
+	}
+
+	// Send response
+	resp := NewResponse(req, true)
+	if err := s.sendMessage(resp); err != nil {
+		return err
+	}
+
+	// Send initialized event
+	event := &InitializedEvent{
+		Event: *NewEvent("initialized"),
+	}
+	return s.sendMessage(event)
+}
+
+// handleSetBreakpoints processes the setBreakpoints request
+func (s *DAPServer) handleSetBreakpoints(req *Request, _ []byte) error {
+	var args SetBreakpointsArguments
+	if err := json.Unmarshal(req.Arguments, &args); err != nil {
+		return err
+	}
+
+	// Clear existing breakpoints for this source
+	delete(s.breakpoints, args.Source.Path)
```

Comment:
Hey very nice job here :), I'm trying to understand all this

If I understand correctly here you delete all breakpoints from DAPServer struct to create the slice of Breakpoints from scratch every time we `handleSetBreakpoints`.

But on  `atBreak(m *Machine) bool` function you are using the breakpoints from Debugger, these breakpoints are always appended but never cleaned

so if you set the breakpoints on line 5 and 7 in a first request

then you set only line 7 (remove breakpoint from 5) I think your DAPServer would have the right breakpoints (only 7) while debugger will have both of them ?

the breakpoint will stay as active ?